### DISCUSSION TITLE
Force Prettier to run on otherwise ignored file

### DISCUSSION THREAD

**Discussion by rschristian** [prettier/prettier#16169] [Q&A]
Sorry if I'm being daft, but either the docs are lacking or this doesn't seem to be supported, which is frustrating.

Often I need to delve into (minifed) `node_modules` to debug some issue, where formatting would be much appreciated. However, it seems to be impossible to force Prettier to write these files, even with a file path explicitly provided, i.e., `prettier --write node_modules/example/index.js`.

With a path to a specific module, the ignore file should be completely ignored, IMO. It's clear that I want that file formatted.

Is there any way of doing this?

**Comment by sphars** [comment]
I'm running into this issue as well. In my situation, my project has a script that will generate a `*.config.js` file for user settings that I do not want committed (filename may differ). So I have added that to my .gitignore:

```
node_modules
*.config.js
```

Understandably, Prettier ignores these files at this point. However, I do want to format this file for readability because the config file is generated as one line and can get pretty long. There are other JS files I do want formatted too.

I have tried negating this rule in my .prettierignore file:

```
# Ignore everything
/**/*.*

# But allow only javascript
!**/*.js
!*.config.js
```

I've also tried adding to my .prettierrc overrides:

```json
{
  "semi": true,
  "tabWidth": 4,
  "useTabs": false,
  "singleQuote": false,
  "trailingComma": "none",
  "bracketSpacing": true,
  "overrides": [
    {
      "files": "*.config.js",
      "options": {
        "tabWidth": 2,
        "printWidth": 80
      }
    }
  ]
}
```

Running `npx prettier --write .` doesn't pick it up, nor does explicitly doing the one file: `npx prettier --write test.config.js`, like OP. I tried a bunch of different patterns in the configs but nothing worked to override the .gitignore.

I noticed that VS Code will format the file if I manually invoke the command "Format Document With..." and select "Prettier". It picks up my local project configuration but doesn't use the .gitignore file.

The only way I can get those `*.config.js` files to be formatted is by removing it from .gitignore.

Alternatively, for my script that creates the `*.config.js` file, I was able to use the prettier API to programmatically format the string to be written. However, the point of this discussion still stands in that I cannot get prettier to format that file directly.

Minimum repro with prettier v3.3.3: https://github.com/sphars/playground/tree/prettier/prettier-16169

**Reply by GrtDev** [reply]
I just ran into this issue myself. It seems like you can set the ignore-path to an empty string and it will work.

```
npx prettier . --write --no-config --ignore-path ''
```

**Reply by flevi29** [reply]
Same problem here. I guess I'll have to live without formatting for now for my specific use case. Or rather forever if I'm waiting on prettier to solve this.

**Comment by peweisberg-fadv** [comment]
`--with-node-modules` gets the job done, in this case.

Though I agree that it makes Prettier seem stupid when it says "No matching files" after I give it the full path to a file that definitely exists.

**Comment by strindhaug** [comment]
I'd personally strongly recommend always making a copy of the file from node_modules to outside of node_modules rather than modifying it directly (and modification includes formatting) inside node_modules. (In fact I'd prefer if it was simple to write protect all of node_modules for any process except for the npm/yarn binary somehow)

Unless you have node_modules committed to the repo (which I've never heard of anyone doing) it's all to easy to change something so you break something inside node_modules and not notice. Or even worse, acciudentally "fixing" a bug by making local changes inside node_modules only for it to break on every other computer when they don't have the accidental modifications. It wouldn't be such a problem if npm/yarn etc. kept some kind of diff themselves and actually removed local changes inside node_modules automatically when running `npm install` or `yarn` or at least warned about local changes; but since they don't you have to selectively downgrade and re-upgrade the packages you worry that you've might changed to reset them; or if you can't remember what files you've touched you have to delete all of node_modules and then take a coffee break while npm downloads "the whole internet" again.

**Reply by rschristian** [reply]
Nonsense; it takes very little responsibility to wield properly and prettier won't (or shouldn't) make any non-formatting changes anyhow. It's very easy to manually diff to see what changes, if any, at the end of the debugging session need to land upstream or continue as patches. Besides, you can't avoid editing `node_modules` in most scenarios, module resolution is what it is and `node_modules` is the magic directory for making that all happen.

If you've done any serious maintenance or debugging work you'd surely know this.

**Reply by strindhaug** [reply]
Sure, if you're very carefully and deliberately making patches to a specific project it's not that hard because presumably you already have made a copy of the original files to diff with (though in that case it's much better to just check out the repo for that project assuming it's open source or a repo you already have access to; and link it in using `yalc publish` and `yalc add`, that way you can see the changes with `git diff` continually and just make a commit when you're done). 

But I was thinking more about randomly navigating into a 3rd party file and make changes to debug stuff; and it's definitely not "nonsense" that it's easy to accidentally make changes you forget or fail to revert properly when doing so, especially if you're ADHD like me or just unexperienced. It's definitely much easier to track what you've changed if you only change files that are indexed by git. Of course making a manual copy still requires you to notice that you _have_ navigated into node_modules but at least having that as a personal policy reduces the chance that you become habituated to making changes directly inside node_modules on purpose.

Really there should be some way of write protecting node_modules somehow without making it inaccessible for the npm command (the most obvious way would have been if npm always ran as a different unix user, so files could be `drwxr-xr-x` and owned by npm), so that you can't accidentally make changes there when navigating to source (option+click, at least with my keyboard setup in vscode) not realising it has jumped into node_modules. That's when I've personally modified things in node_modules by accident most often... fortunately I tend not to habitually close files too often so the changed files are usually still in an open tab; though I have had to completely delete node_modules a few times, when I realise I must have changed something in a local copy days ago and the same branch suddenly doesn't behave the same on a different checkout of the repo. And that happened much more often before I decided to never deliberately modify node_modules. (Sorry I'm getting off topic, as I said ADHD 😅 )
